tmt/steps/provision/__init__.py

```diff
@@ -138,15 +152,23 @@ def go(self):
         try:
             for phase in self.phases(classes=(Action, ProvisionPlugin)):
                 try:
-                    phase.go()
-                    if isinstance(phase, ProvisionPlugin):
-                        phase.guest().details()
+                    if isinstance(phase, Action):
+                        phase.go()
+
+                    elif isinstance(phase, ProvisionPlugin):
+                        phase.go()
+
+                        guest = phase.guest()
+                        if guest:
+                            guest.details()
```

Comment:
Could this part be simplified to this?
```suggestion
                    phase.go()
                    if isinstance(phase, ProvisionPlugin):
                        guest = phase.guest()
                        if guest:
                            guest.details()
```

Reply:
It was like this :) But mypy was unhappy about `phase` maybe not having `go()` method, and I rewrote conditions because 1. mypy is limited, it's not allowed to see everything, and while the core was correct, I gave it the benefit of doubt, and 2. it's similar to other steps that check for their plugin and action in their own branches and raise an error if anything else slips in.

Reply:
I see, so it boils down to `self.phases()` not being annotated properly (or mypy not handling it properly). Looking at some newer Python features, would `TypeVarTuple` (https://peps.python.org/pep-0646/) fix the `phases()` issue, if I understand it correctly?

`TypeVarTuple` is already available in `typing_extensions` module today. However, mypy doesn't support it yet. :)

Reply:
Actually, I thought `self.phases()` *was* annotated correctly, that's why it was a bit weird. My impression was that it was rather mypy who had troubles because the language server in my Code IDE was reporting the expected type of `phase`, `Union[Action, ProvisionPlugin]`, while mypy was stuck with `tmt.utils.Common`.

Reply:
Not sure what's correct then, I'm getting the same results as you in VS Code. No point of discussing it anyways, since mypy doesn't even support it. :)